framework/src/tinyalsa/tinyalsa.c

```diff
@@ -532,51 +533,74 @@ int pcm_readi(struct pcm *pcm, void *data, unsigned int frame_count)
 	nbytes = pcm_frames_to_bytes(pcm, frame_count);
 
 #ifdef CONFIG_AUDIO_DRIVER_SPECIFIC_BUFFERS
-	if (nbytes < pcm->config.period_size) {
-		return -EINVAL;
-	}
 	bufdesc.numbytes = pcm->config.period_size;
 #else
-	if (nbytes < CONFIG_AUDIO_BUFFER_NUMBYTES) {
-		return -EINVAL;
-	}
 	bufdesc.numbytes = CONFIG_AUDIO_BUFFER_NUMBYTES;
 #endif
 #ifdef CONFIG_AUDIO_MULTI_SESSION
 	bufdesc.session = pcm->session;
 #endif
 
-	/* If device is not yet started, start now! */
-	if ((!pcm->running) && (pcm_start(pcm) < 0)) {
-		return -errno;
-	} else {
-		pcm->running = 1;
+	/* If no pending buffers ,then enqueue some buffers*/
+	if (pcm->enqueuedBytes==0) {
+		/* If the device is opened for read and Our buffers are not enqued. Enque them now. */
+#ifdef CONFIG_AUDIO_DRIVER_SPECIFIC_BUFFERS
```

Comment:
As I commented, we don't have to check CONFIG_AUDIO_DRIVER_SPECIFIC_BUFFERS here.
pcm->config.period_count knows number of buffer

Reply:
We have to consider the case when user gives incomplete config in set_config call of tinyalsa. If the user overwrites the period_size and period_count with some random values, then readi call would not succeed.